[AST/Sema] TypeWrappers: Fix a couple of issues with attribute inference from protocols

xedin opened the pull request:
- ASTPrinter: Feature coverage should include type wrapped types as well (directly and indirectly depending on the feature) otherwise compiler is going to produce a broken swiftinterface file.

- TypeChecker: Synthesis should produce a `public` implementation of `$Storage` and `$storage` if protocol requirements are public.

Resolves: rdar://103270262

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ASTPrinter] Expand type wrapper feature coverage

`usesFeatureTypeWrappers` should return `true` for all of
the declarations that use the feature directly or indirectly
e.g. type wrapper types, type wrapped types, and types with
conformance to a type wrapped protocol (with or without wrapper
attribute inference).

xedin committed:
[TypeChecker] TypeWrappers: Set access level for $Storage/$storage based on protocol conformances

If a type conforms to a protocol that has a public `$Storage`/`$storage`
requirement the synthesis should produce a public `$Storage`/`$storage`
implementation as well to match the requirement.

Resolves: rdar://103270262

xedin committed:
[AST] TypeWrappers/NFC: Move local $storage handling code to TypeWrapper.cpp

tshortli commented on test/ModuleInterface/type_wrappers.swift:
```diff
@@ -0,0 +1,123 @@
+// CHECK:   public init(storageWrapper: TypeWrappers.Wrapper<TypeWrappers.ClassWithProtoocol, TypeWrappers.ClassWithProtoocol.$Storage>)
+// CHECK: }
+
+@Wrapper
```
Is the explicit `@Wrapper` intentional here? I thought this test case might be testing the same thing as above, but with a `class` instead of a `struct`.

xedin replied:
It is intentional to make sure that there is no clash there.

xedin replied:
I actually realized that there is no attribute here although I think there should be, I'm currently working on fixing that.

xedin replied:
Fixed!

tshortli reviewed (approved):
LGTM!

xedin committed:
[AST/Sema] TypeWrappers: Augment `GetTypeWrapper` request to produce more info

The request is updated to return attribute, wrapper declaration,
the declaration its attached to and whether or not it has been
inferred (opposite to being declared directly).

xedin committed:
[ASTPrinter] TypeWrappers: print inferred attributes

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please test Windows platform
